File: backend/rutas/estadisticas.py
from fastapi import APIRouter, Depends
from sqlalchemy.orm import Session
from sqlalchemy import func
from datetime import datetime, timedelta
from dependencias import obtener_db, obtener_usuario_actual, requerir_admin
from modelos.tarea import Tarea
from modelos.usuario import Usuario
from esquemas.estadistica import RankingUsuariosSalida, ItemRankingUsuario, ResumenPropio, KpisUsuario, KpisUsuariosSalida

router = APIRouter(prefix="/estadisticas", tags=["estadisticas"])

PESO = {"alta": 3, "media": 2, "baja": 1}


@router.get("/ranking-usuarios", response_model=RankingUsuariosSalida)
def ranking_usuarios(db: Session = Depends(obtener_db), _: Usuario = Depends(requerir_admin)):
    tareas = (
        db.query(Tarea)
        .filter(Tarea.estado == "finalizada", Tarea.asignado_a.isnot(None))
        .all()
    )
    acumulado: dict[int, dict] = {}
    for t in tareas:
        uid = t.asignado_a
        if uid not in acumulado:
            u = db.query(Usuario).filter(Usuario.id == uid).first()
            acumulado[uid] = {"nombre": u.nombre if u else "", "apellido": u.apellido if u else "", "puntos": 0, "tareas_finalizadas": 0}
        acumulado[uid]["puntos"] += PESO.get(t.criticidad, 1)
        acumulado[uid]["tareas_finalizadas"] += 1

    ranking = [
        ItemRankingUsuario(usuario_id=uid, **datos)
        for uid, datos in sorted(acumulado.items(), key=lambda x: x[1]["puntos"], reverse=True)
    ]
    return RankingUsuariosSalida(ranking=ranking)


@router.get("/mi-resumen", response_model=ResumenPropio)
def mi_resumen(db: Session = Depends(obtener_db), usuario: Usuario = Depends(obtener_usuario_actual)):
    corte_7 = datetime.utcnow() - timedelta(days=7)

    en_proceso = db.query(func.count(Tarea.id)).filter(
        Tarea.estado == "en_proceso", Tarea.asignado_a == usuario.id
    ).scalar() or 0

    finalizadas_7 = db.query(func.count(Tarea.id)).filter(
        Tarea.estado == "finalizada",
        Tarea.asignado_a == usuario.id,
        Tarea.fecha_finalizacion >= corte_7,
    ).scalar() or 0

    finalizadas_total = db.query(func.count(Tarea.id)).filter(
        Tarea.estado == "finalizada", Tarea.asignado_a == usuario.id
    ).scalar() or 0

    ultimas = (
        db.query(Tarea)
        .filter(Tarea.estado == "finalizada", Tarea.asignado_a == usuario.id)
        .order_by(Tarea.fecha_finalizacion.desc())
        .limit(5)
        .all()
    )

    return ResumenPropio(
        en_proceso=en_proceso,
        finalizadas_7_dias=finalizadas_7,
        finalizadas_total=finalizadas_total,
        ultimas_finalizadas=[{"id": t.id, "titulo": t.titulo, "fecha": str(t.fecha_finalizacion)} for t in ultimas],
    )


@router.get("/kpis-usuarios", response_model=KpisUsuariosSalida)
def kpis_usuarios(db: Session = Depends(obtener_db), _: Usuario = Depends(requerir_admin)):
    usuarios = db.query(Usuario).filter(Usuario.activo == True).all()
    resultado: list[KpisUsuario] = []

    for u in usuarios:
        tareas = db.query(Tarea).filter(Tarea.asignado_a == u.id).all()
        if not tareas:
            continue

        finalizadas = [t for t in tareas if t.estado == "finalizada"]
        en_proceso = [t for t in tareas if t.estado == "en_proceso"]
        por_hacer = [t for t in tareas if t.estado == "por_hacer"]

        # Puntos
        puntos = sum(PESO.get(t.criticidad, 1) for t in finalizadas)

        # Distribución criticidad (todas las tareas asignadas)
        alta = sum(1 for t in tareas if t.criticidad == "alta")
        media = sum(1 for t in tareas if t.criticidad == "media")
        baja = sum(1 for t in tareas if t.criticidad == "baja")
        total = len(tareas)
        pct_alta = round((alta / total) * 100, 1) if total else 0.0

        # Promedio respuesta: creacion → inicio (tareas que tienen fecha_inicio)
        tiempos_respuesta = [
            (t.fecha_inicio - t.fecha_creacion).total_seconds() / 3600
            for t in tareas
            if t.fecha_inicio and t.fecha_creacion and t.fecha_inicio >= t.fecha_creacion
        ]
        prom_respuesta = round(sum(tiempos_respuesta) / len(tiempos_respuesta), 1) if tiempos_respuesta else None

        # Promedio resolución: inicio → finalización (tareas finalizadas con ambas fechas)
        tiempos_resolucion = [
            (t.fecha_finalizacion - t.fecha_inicio).total_seconds() / 3600
            for t in finalizadas
            if t.fecha_inicio and t.fecha_finalizacion and t.fecha_finalizacion >= t.fecha_inicio
        ]
        prom_resolucion = round(sum(tiempos_resolucion) / len(tiempos_resolucion), 1) if tiempos_resolucion else None

        resultado.append(KpisUsuario(
            usuario_id=u.id,
            nombre=u.nombre,
            apellido=u.apellido,
            puntos=puntos,
            tareas_finalizadas=len(finalizadas),
            tareas_en_proceso=len(en_proceso),
            tareas_por_hacer=len(por_hacer),
            promedio_respuesta_horas=prom_respuesta,
            promedio_resolucion_horas=prom_resolucion,
            tareas_alta=alta,
            tareas_media=media,
            tareas_baja=baja,
            pct_alta=pct_alta,
        ))

    resultado.sort(key=lambda x: x.puntos, reverse=True)
    return KpisUsuariosSalida(usuarios=resultado)

File: backend-node/src/routes/estadisticas.js
import { Router } from 'express';
import { prisma } from '../db.js';
import { obtenerUsuarioActual, requerirSuperAdminOSuperior } from '../middleware/auth.js';

const router = Router();

const PESO = { alta: 3, media: 2, baja: 1 };

// GET /api/estadisticas/ranking-usuarios
router.get('/ranking-usuarios', requerirSuperAdminOSuperior, async (req, res) => {
  const tareas = await prisma.tarea.findMany({
    where: { estado: 'finalizada', asignados: { some: {} } },
    include: { asignados: true },
  });

  const acumulado = {};

  for (const t of tareas) {
    for (const u of t.asignados) {
      const uid = u.id;
      if (!acumulado[uid]) {
        acumulado[uid] = { usuario_id: uid, nombre: u.nombre || '', apellido: u.apellido || '', puntos: 0, tareas_finalizadas: 0 };
      }
      acumulado[uid].puntos += PESO[t.criticidad] ?? 1;
      acumulado[uid].tareas_finalizadas += 1;
    }
  }

  const ranking = Object.values(acumulado).sort((a, b) => b.puntos - a.puntos);
  return res.json({ ranking });
});

// GET /api/estadisticas/mi-resumen
router.get('/mi-resumen', obtenerUsuarioActual, async (req, res) => {
  const uid = req.usuario.id;
  const corte7 = new Date(Date.now() - 7 * 24 * 60 * 60 * 1000);

  const [en_proceso, finalizadas_7_dias, finalizadas_total, ultimas] = await Promise.all([
    prisma.tarea.count({ where: { estado: 'en_proceso', asignados: { some: { id: uid } } } }),
    prisma.tarea.count({ where: { estado: 'finalizada', asignados: { some: { id: uid } }, fecha_finalizacion: { gte: corte7 } } }),
    prisma.tarea.count({ where: { estado: 'finalizada', asignados: { some: { id: uid } } } }),
    prisma.tarea.findMany({
      where: { estado: 'finalizada', asignados: { some: { id: uid } } },
      orderBy: { fecha_finalizacion: 'desc' },
      take: 5,
      select: { id: true, titulo: true, fecha_finalizacion: true },
    }),
  ]);

  return res.json({
    en_proceso,
    finalizadas_7_dias,
    finalizadas_total,
    ultimas_finalizadas: ultimas.map(t => ({ id: t.id, titulo: t.titulo, fecha: String(t.fecha_finalizacion) })),
  });
});

// GET /api/estadisticas/kpis-usuarios
router.get('/kpis-usuarios', requerirSuperAdminOSuperior, async (req, res) => {
  const usuarios = await prisma.usuario.findMany({ where: { activo: true } });
  const resultado = [];

  for (const u of usuarios) {
    const tareas = await prisma.tarea.findMany({ where: { asignados: { some: { id: u.id } } } });
    if (!tareas.length) continue;

    const finalizadas = tareas.filter(t => t.estado === 'finalizada');
    const en_proceso = tareas.filter(t => t.estado === 'en_proceso');
    const por_hacer = tareas.filter(t => t.estado === 'por_hacer');

    const puntos = finalizadas.reduce((acc, t) => acc + (PESO[t.criticidad] ?? 1), 0);

    const alta = tareas.filter(t => t.criticidad === 'alta').length;
    const media = tareas.filter(t => t.criticidad === 'media').length;
    const baja = tareas.filter(t => t.criticidad === 'baja').length;
    const total = tareas.length;
    const pct_alta = total ? Math.round((alta / total) * 1000) / 10 : 0;

    const tiempos_respuesta = tareas
      .filter(t => t.fecha_inicio && t.fecha_creacion && t.fecha_inicio >= t.fecha_creacion)
      .map(t => (t.fecha_inicio - t.fecha_creacion) / 3600000);
    const prom_respuesta = tiempos_respuesta.length
      ? Math.round((tiempos_respuesta.reduce((a, b) => a + b, 0) / tiempos_respuesta.length) * 10) / 10
      : null;

    const tiempos_resolucion = finalizadas
      .filter(t => t.fecha_inicio && t.fecha_finalizacion && t.fecha_finalizacion >= t.fecha_inicio)
      .map(t => (t.fecha_finalizacion - t.fecha_inicio) / 3600000);
    const prom_resolucion = tiempos_resolucion.length
      ? Math.round((tiempos_resolucion.reduce((a, b) => a + b, 0) / tiempos_resolucion.length) * 10) / 10
      : null;

    resultado.push({
      usuario_id: u.id,
      nombre: u.nombre,
      apellido: u.apellido,
      puntos,
      tareas_finalizadas: finalizadas.length,
      tareas_en_proceso: en_proceso.length,
      tareas_por_hacer: por_hacer.length,
      promedio_respuesta_horas: prom_respuesta,
      promedio_resolucion_horas: prom_resolucion,
      tareas_alta: alta,
      tareas_media: media,
      tareas_baja: baja,
      pct_alta,
    });
  }

  resultado.sort((a, b) => b.puntos - a.puntos);
  return res.json({ usuarios: resultado });
});

export default router;
